beacon-chain/blockchain/core.go

```diff
@@ -249,3 +256,255 @@ func (b *BeaconChain) getAttestersProposer(seed common.Hash) ([]int, int, error)
 	}
 	return indices[:int(attesterCount)], indices[len(indices)-1], nil
 }
+
+// hasVoted checks if the attester has voted by looking at the bitfield.
+func hasVoted(bitfields []byte, attesterBlock int, attesterFieldIndex int) bool {
+	voted := false
+
+	fields := bitfields[attesterBlock-1]
+	attesterField := fields >> (8 - uint(attesterFieldIndex))
+	if attesterField%2 != 0 {
+		voted = true
+	}
+
+	return voted
+}
+
+// applyRewardAndPenalty applies the appropriate rewards and penalties according to
+// whether the attester has voted or not.
+func (b *BeaconChain) applyRewardAndPenalty(index int, voted bool) error {
+	defer b.lock.Unlock()
+	b.lock.Lock()
+
+	if voted {
+		b.state.CrystallizedState.ActiveValidators[index].Balance += params.AttesterReward
+	} else {
+		// TODO : Change this when penalties are specified for not voting
+		b.state.CrystallizedState.ActiveValidators[index].Balance -= params.AttesterReward
+	}
+
+	return b.persist()
+}
+
+// resetAttesterBitfields resets the attester bitfields in the ActiveState to zero.
+func (b *BeaconChain) resetAttesterBitfields() error {
+
+	length := int(len(b.state.CrystallizedState.ActiveValidators) / 8)
+	if len(b.state.CrystallizedState.ActiveValidators)%8 != 0 {
+		length += 1
+	}
+
+	defer b.lock.Unlock()
+	b.lock.Lock()
+
+	newbitfields := make([]byte, length)
+	b.state.ActiveState.AttesterBitfields = newbitfields
+
+	return b.persist()
+}
+
+// resetTotalDeposit clears and resets the total attester deposit to zero.
+func (b *BeaconChain) resetTotalDeposit() error {
+	defer b.lock.Unlock()
+	b.lock.Lock()
+	b.state.ActiveState.TotalAttesterDeposits = 0
+
+	return b.persist()
+}
+
+// setJustifiedEpoch sets the justified epoch during an epoch transition.
+func (b *BeaconChain) setJustifiedEpoch() error {
```

Comment:
Can you choose a better method name? It sounds weird to set something but not provide any arguments/parameters.